pkg/apps/controller/util/util.go

```diff
@@ -0,0 +1,901 @@
+	for _, trigger := range config.Spec.Triggers {
+		if trigger.Type == appsapi.DeploymentTriggerOnImageChange {
+			return true
+		}
+	}
+	return false
+}
+
+// HasTrigger returns whether the provided deployment configuration has any trigger
+// defined or not.
+func HasTrigger(config *appsapi.DeploymentConfig) bool {
+	return HasChangeTrigger(config) || HasImageChangeTrigger(config)
+}
+
+// HasLastTriggeredImage returns whether all image change triggers in provided deployment
+// configuration has the lastTriggerImage field set (iow. all images were updated for
+// them). Returns false if deployment configuration has no image change trigger defined.
+func HasLastTriggeredImage(config *appsapi.DeploymentConfig) bool {
+	hasImageTrigger := false
+	for _, trigger := range config.Spec.Triggers {
+		if trigger.Type == appsapi.DeploymentTriggerOnImageChange {
+			hasImageTrigger = true
+			if len(trigger.ImageChangeParams.LastTriggeredImage) == 0 {
+				return false
+			}
+		}
+	}
+	return hasImageTrigger
+}
+
+// IsInitialDeployment returns whether the deployment configuration is the first version
+// of this configuration.
+func IsInitialDeployment(config *appsapi.DeploymentConfig) bool {
+	return config.Status.LatestVersion == 0
+}
+
+// RecordConfigChangeCause sets a deployment config cause for config change.
+func RecordConfigChangeCause(config *appsapi.DeploymentConfig) {
+	config.Status.Details = &appsapi.DeploymentDetails{
+		Causes: []appsapi.DeploymentCause{
+			{
+				Type: appsapi.DeploymentTriggerOnConfigChange,
+			},
+		},
+		Message: "config change",
+	}
+}
+
+// RecordImageChangeCauses sets a deployment config cause for image change. It
+// takes a list of changed images and record an cause for each image.
+func RecordImageChangeCauses(config *appsapi.DeploymentConfig, imageNames []string) {
+	config.Status.Details = &appsapi.DeploymentDetails{
+		Message: "image change",
+	}
+	for _, imageName := range imageNames {
+		config.Status.Details.Causes = append(config.Status.Details.Causes, appsapi.DeploymentCause{
+			Type:         appsapi.DeploymentTriggerOnImageChange,
+			ImageTrigger: &appsapi.DeploymentCauseImageTrigger{From: api.ObjectReference{Kind: "DockerImage", Name: imageName}},
+		})
+	}
+}
+
+func CopyApiResourcesToV1Resources(in *api.ResourceRequirements) v1.ResourceRequirements {
+	in = in.DeepCopy()
+	out := v1.ResourceRequirements{}
+	if err := kapiv1.Convert_core_ResourceRequirements_To_v1_ResourceRequirements(in, &out, nil); err != nil {
+		panic(err)
+	}
+	return out
+}
+
+func CopyApiEnvVarToV1EnvVar(in []api.EnvVar) []v1.EnvVar {
+	out := make([]v1.EnvVar, len(in))
+	for i := range in {
+		if err := kapiv1.Convert_core_EnvVar_To_v1_EnvVar(in[i].DeepCopy(), &out[i], nil); err != nil {
+			panic(err)
+		}
+	}
+	return out
+}
+
+// HasUpdatedImages indicates if the deployment configuration images were updated.
+func HasUpdatedImages(dc *appsapi.DeploymentConfig, rc *v1.ReplicationController) (bool, []string) {
```

Comment:
seems like we are mixing clients - inside one component we should use internal client for both DC and RC, right?

Reply:
what I meant was that whoever is calling this (our controller) is mixing internal client for DCs and external one for RCs - but that's a followup discussion

Reply:
@tnozicka 
```
func triggerActivated(config *appsapi.DeploymentConfig, latestExists bool, latestDeployment *v1.ReplicationController) (bool, bool, error) {
```

from deploymentconfig controller... we can fix that but the DC originates from informer this time...